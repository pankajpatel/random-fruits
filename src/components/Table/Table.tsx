import styled from "styled-components";

type RowObject = Record<string, any>;

export interface TableCell {
  label: string;
  key: string;
  render?: (row: any) => JSX.Element;
  align?: "left" | "right";
  sortable?: boolean;
}

interface TableProps {
  rows: Array<RowObject>;
  cells: TableCell[];
  onRowClick?: (row: RowObject) => void;
  onHeaderClick?: (key: string) => void;
}

const StyledTable = styled.table`
  border-collapse: collapse;
`;

const Row = styled.tr`
  cursor: pointer;
  font-size: 1.1rem;

  &:nth-child(even) {
    background-color: #fafafa;
  }

  &:hover {
    background-color: #eee;
  }
`;

const Cell = styled.td<{ align: "left" | "right" }>`
  padding: 0.5rem 0.75rem;
  text-align: ${(props) => props.align ?? "left"};
`;

const HeaderCell = styled.th<{ align: "left" | "right" }>`
  padding: 0.75rem;
  text-align: ${(props) => props.align ?? "left"};
`;

export const Table = ({
  rows,
  cells,
  onRowClick = () => {},
  onHeaderClick = () => {},
}: TableProps): JSX.Element => {

  return (
    <StyledTable>
      <thead>
        <tr>
          {cells.map(({ label, align = "left", key }) => (
            <HeaderCell align={align} key={label} onClick={() => onHeaderClick(key)}>
              {label}
            </HeaderCell>
          ))}
        </tr>
      </thead>
      <tbody>
        {rows.map((row: RowObject) => (
          <Row onClick={() => onRowClick(row)} key={row.id} title="Click open Payment Details">
            {cells.map(({ label, key, render, align = "left" }: TableCell) => {
              if (typeof render === "function") {
                return (
                  <Cell align={align} key={label}>
                    {render(row)}
                  </Cell>
                );
              }
              if (key) {
                return (
                  <Cell align={align} key={label}>
                    {row[key]}
                  </Cell>
                );
              }
              return (
                <Cell align={align} key={label}>
                  ' '
                </Cell>
              );
            })}
          </Row>
        ))}
      </tbody>
    </StyledTable>
  );
};
