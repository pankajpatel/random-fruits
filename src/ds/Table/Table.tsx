import { ReactNode } from 'react';
import styled from 'styled-components';

export interface TableCell<T extends unknown> {
  label: ReactNode;
  key: keyof T;
  render?: (row: T) => ReactNode;
  align?: 'left' | 'right';
  sortable?: boolean;
  maxWidth?: string;
}

interface TableProps<T> {
  rows: Array<T>;
  cells: TableCell<T>[];
  onRowClick?: (row: T) => void;
  onHeaderClick?: (key: keyof T) => void;
}

const StyledTable = styled.table`
  border-collapse: collapse;
  width: 100%;
`;

const Row = styled.tr`
  cursor: pointer;
  font-size: 1.1rem;

  &:nth-child(even) {
    background-color: #fafafa;
  }

  &:hover {
    background-color: #eee;
  }
`;

const Cell = styled.td<{ align: 'left' | 'right' }>`
  padding: 0.75rem 0.5rem;
  text-align: ${(props) => props.align ?? 'left'};
`;

const HeaderCell = styled.th<{ align: 'left' | 'right'; clickable: boolean }>`
  padding: 0.75rem 0.5rem;
  text-align: ${(props) => props.align ?? 'left'};
  ${(props) =>
    props.clickable &&
    `
    cursor: pointer;
    &:hover { background-color: #eee; }
  `};
`;

const NOOP = () => {};

export const Table = <T extends unknown>({
  rows,
  cells,
  onRowClick = NOOP,
  onHeaderClick,
}: TableProps<T>): JSX.Element => (
  <StyledTable>
    <thead>
      <tr>
        {cells.map(({ label, align = 'left', key, sortable }) => (
          <HeaderCell
            key={key as string}
            align={align}
            clickable={Boolean(sortable)}
            onClick={() =>
              (sortable && onHeaderClick ? onHeaderClick : NOOP)(key)
            }
          >
            {label}
            {sortable && <small>â‡µ</small>}
          </HeaderCell>
        ))}
      </tr>
    </thead>
    <tbody>
      {rows.map((row: T, index: number) => (
        <Row
          key={index}
          onClick={() => onRowClick(row)}
          title="Click open Payment Details"
        >
          {cells.map(
            ({
              label,
              key,
              render,
              align = 'left',
            }: TableCell<T>): ReactNode => {
              const cellKey = key as string;
              if (typeof render === 'function') {
                return (
                  <Cell align={align} key={cellKey}>
                    {render(row)}
                  </Cell>
                );
              }
              if (key) {
                return (
                  <Cell align={align} key={cellKey}>
                    <>{row[key]}</>
                  </Cell>
                );
              }
              return (
                <Cell align={align} key={cellKey}>
                  ' '
                </Cell>
              );
            }
          )}
        </Row>
      ))}
    </tbody>
  </StyledTable>
);
