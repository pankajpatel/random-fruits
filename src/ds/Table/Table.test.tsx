import { render } from "@testing-library/react";
import { Table, TableCell } from "./Table";

type MockRow = {
  status: string;
  id: string;
  created: number;
  customer_name: string;
  amount: number;
  installmentsCount: number;
};

const mockConfig: TableCell<MockRow>[] = [
  {
    label: "Customer Name",
    key: "customer_name",
  },
  {
    label: "Created On",
    key: "created",
    render: (row: MockRow) => <>{new Date(row.created).toLocaleDateString()}</>,
  },
];

const mockRow = {
  status: "ready",
  id: "payment_i2NJhL",
  created: 1649023200000,
  customer_name: "Ã‰mile Zola",
  amount: 356050,
  installmentsCount: 10,
};

describe("Table", () => {
  it("should render without crashing", () => {
    const { container } = render(
      <Table cells={mockConfig} rows={[mockRow]} />
    );
    expect(container).toMatchSnapshot();
  });
});
