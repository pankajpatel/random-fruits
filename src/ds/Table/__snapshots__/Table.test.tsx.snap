// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should render without crashing 1`] = `
.c0 {
  border-collapse: collapse;
  width: 100%;
}

.c2 {
  cursor: pointer;
  font-size: 1.1rem;
}

.c2:nth-child(even) {
  background-color: #fafafa;
}

.c2:hover {
  background-color: #eee;
}

.c3 {
  padding: 0.75rem 0.5rem;
  text-align: left;
}

.c1 {
  padding: 0.75rem 0.5rem;
  text-align: left;
}

<div>
  <table
    class="c0"
  >
    <thead>
      <tr>
        <th
          class="c1"
        >
          Customer Name
        </th>
        <th
          class="c1"
        >
          Created On
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="c2"
        title="Click open Payment Details"
      >
        <td
          class="c3"
        >
          Ã‰mile Zola
        </td>
        <td
          class="c3"
        >
          04/04/2022
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
